traversal/expression/Operation.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.traversal.expression;
+
+import com.vaticle.typedb.common.collection.Triple;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.encoding.Encoding;
+import com.vaticle.typedb.core.graph.vertex.ValueHolder;
+import com.vaticle.typedb.core.graph.vertex.impl.ValueVertexImpl;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.DUPLICATE_EXPRESSION_DEFINITION;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.UNIMPLEMENTED;
+
+public abstract class Operation<T1, T2> implements ExpressionFactory.OperationFactory<T1, T2> {
+    private static Map<Triple<java.lang.String, Encoding.ValueType<?>, Encoding.ValueType<?>>,ExpressionFactory.OperationFactory> registry = null;
```

Comment:
Had to use an unparametrised OperationFactory here. 
The operand types vary from signature to signature, So I had trouble getting the `build` call in  `ExpressionFactory.instantiateOperation` to compile. I might be able to do parametrise that function, but that just pushes the problem to the caller.